File: backend/src/igab/api/v1/schemas/emergency_fund.py
"""The emergency fund picker (`api/v1/emergency_fund.py`)."""

import uuid
from decimal import Decimal

from pydantic import ConfigDict, Field

from igab.api.v1.schemas.base import ApiModel
from igab.api.v1.schemas.report import EmergencyFundOut
from igab.repositories.category_filters import SavingsMode


class FundAccountCandidateOut(ApiModel):
    """A live, open off-budget account the fund may count
    (`txn_filters.EMERGENCY_FUND_ACCOUNT_CANDIDATE`)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    balance: Decimal
    #: False means choosing it also turns Counts as savings on.
    counts_as_savings: bool
    #: Counted by the fund now.
    member: bool


class EmergencyFundPickerOut(ApiModel):
    fund: EmergencyFundOut
    account_candidates: list[FundAccountCandidateOut]


class FundExternalChoice(ApiModel):
    declared: bool
    #: Canonical decimal string or null — "I have this covered" without a figure
    #: is a complete answer, never zero.
    amount: Decimal | None = Field(default=None, ge=0)
    note: str | None = None


class EmergencyFundChoiceRequest(ApiModel):
    """The picker's one save. Envelopes are a diff against the Emergency fund
    tag's checklist as loaded; accounts are the full chosen set."""

    add_categories: list[uuid.UUID] = []
    remove_categories: list[uuid.UUID] = []
    savings_modes: dict[uuid.UUID, SavingsMode | None] = {}
    account_ids: list[uuid.UUID]
    external: FundExternalChoice

File: frontend/src/api/emergencyFund.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'
import { ROOT } from './queryKeys'
import { invalidateAfterTagChange } from './invalidateAfterTagChange'
import type { EmergencyFund, SavingsMode } from '../types'

/** An account the emergency fund may count: a live, open off-budget asset
 *  (`txn_filters.EMERGENCY_FUND_ACCOUNT_CANDIDATE`). */
export interface FundAccountCandidate {
  id: string
  name: string
  balance: number
  /** False: choosing it also turns Counts as savings on, in the same save. */
  counts_as_savings: boolean
  /** Counted by the fund now. */
  member: boolean
}

/** GET /{budget}/emergency-fund — what the fund counts, and what it could. */
export interface EmergencyFundPicker {
  fund: EmergencyFund
  account_candidates: FundAccountCandidate[]
}

/** The picker's one save (PUT /{budget}/emergency-fund). Envelopes are a diff
 *  against the Emergency fund tag's checklist; accounts are the full set. */
export interface EmergencyFundChoice {
  add_categories: string[]
  remove_categories: string[]
  savings_modes: Record<string, SavingsMode | null>
  account_ids: string[]
  external: {
    declared: boolean
    /** Canonical decimal string, or null for "I have this covered". */
    amount: string | null
    note: string | null
  }
}

export function useEmergencyFund(budgetId: string | null) {
  return useQuery({
    queryKey: [ROOT.emergencyFund, budgetId],
    queryFn: () =>
      apiClient.get<EmergencyFundPicker>(`/${budgetId}/emergency-fund`).then((r) => r.data),
    enabled: !!budgetId,
  })
}

/** One save, one undo: envelopes, modes, accounts and the declared amount. */
export function useSetEmergencyFund(budgetId: string | null) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (choice: EmergencyFundChoice) =>
      apiClient.put<EmergencyFundPicker>(`/${budgetId}/emergency-fund`, choice).then((r) => r.data),
    onSuccess: async (picker) => {
      qc.setQueryData([ROOT.emergencyFund, budgetId], picker)
      // Tags, categories, every report and Guide figure that reads the fund
      // (dashboard, essentials, coverage, savings rate, signals, checkup,
      // sizer) — and the accounts whose flags moved.
      await Promise.all([
        invalidateAfterTagChange(qc, budgetId),
        qc.invalidateQueries({ queryKey: [ROOT.accounts, budgetId] }),
      ])
    },
  })
}
